src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py

```diff
@@ -0,0 +1,139 @@
+    "active": boolean,
+    "pileupSizeGB": integer, current size of the pileup in GB (integer)
+    "rulesList: list of strings (rules) used to lock the pileup id
+}
+
+The data flow should be done via list of objects, e.g.
+[<pileup object 1>, <pileup object 2>, ..., <pileup object n>]
+"""
+
+# system modules
+import time
+import json
+
+# WMCore modules
+from WMCore.MicroService.Tools.Common import getMSLogger
+
+
+def gmtimeSeconds():
+    """
+    Return GMT time in seconds
+    """
+    return int(time.mktime(time.gmtime()))
+
+
+class MSPileupObj(object):
+    """
+    MSPileupObj defines MSPileup data stucture
+    """
+    def __init__(self, pdict, verbose=None, logger=None):
+        self.logger = getMSLogger(verbose, logger)
+
+        self.pileupName = pdict.get('pileupName', '')
+        self.pileupType = pdict.get('pileupType', 'classic')
+        self.insertTime = pdict.get('insertTime', gmtimeSeconds())
+        self.lastUpdateTime = pdict.get('lastUpdateTime', gmtimeSeconds())
+        self.expectedRSEs = pdict.get('expectedRSEs', [])
+        self.fullReplicas = pdict.get('fullReplicas', 0)
+        self.campaigns = pdict.get('campaigns', [])
+        self.containerFraction = pdict.get('containerFraction', 0.0)
+        self.replicationGrouping = pdict.get('replicationGrouping', "")
+        self.activatedOn = pdict.get('activatedOn', gmtimeSeconds())
+        self.deactivatedOn = pdict.get('deactivatedOn', gmtimeSeconds())
+        self.active = pdict.get('active', False)
+        self.pileupSize = pdict.get('pileupSize', 0)
+        self.ruleList = pdict.get('ruleList', [])
+
+        self.data = {
+            'pileupName': self.pileupName,
+            'pileupType': self.pileupType,
+            'insertTime': self.insertTime,
+            'lastUpdateTime': self.lastUpdateTime,
+            'expectedRSEs': self.expectedRSEs,
+            'fullReplicas': self.fullReplicas,
+            'campaigns': self.campaigns,
+            'containerFraction': self.containerFraction,
+            'replicationGrouping': self.replicationGrouping,
+            'activatedOn': self.activatedOn,
+            'deactivatedOn': self.deactivatedOn,
+            'active': self.active,
+            'pileupSize': self.pileupSize,
+            'ruleList': self.ruleList}
+        if not self.validate(self.data):
+            msg = 'MSPileup input is invalid'
+            raise Exception(msg)
+
+    def __str__(self):
+        """
+        Return human readable representation of pileup data
+        """
+        return json.dumps(self.data, indent=4)
+
+    def validate(self, pdict):
+        """
+        Data validator according to its schema
+
+        :param pdict: input data dictionary
+        :return: boolean, result of validation
+        """
+        docSchema = self.schema()
+        for key, val in pdict.items():
+            msg = f"Failed to validate: {key}, expect data-type string got type=" + str(type(val))
```

Comment:
I had a previous comment on this function pointing to two issues:
1) this error msg is wrong. Not all attributes are meant to be of string type
2) and how to make the complexity of this code cheaper (O(n) instead of O(n^2))

Please check your notifications and consider those as well. I can no longer see them when reviewing this PR, but I know I commented on it.

Reply:
done